zerver/actions/custom_profile_fields.py

```diff
@@ -208,3 +214,34 @@ def check_remove_custom_profile_field_value(user_profile: UserProfile, field_id:
         raise JsonableError(_("Field id {id} not found.").format(id=field_id))
     except CustomProfileFieldValue.DoesNotExist:
         pass
+
+
+def validate_user_can_update_custom_profile_data(
+    user_profile: UserProfile,
+    data: List[ProfileDataElementUpdateDict],
+) -> None:
+    with transaction.atomic():
+        for custom_profile_field in data:
+            try:
+                field, created = CustomProfileField.objects.get_or_create(
+                    id=custom_profile_field["id"]
+                )
+                if not field.editable_by_user and not user_profile.is_realm_admin:
+                    raise JsonableError(_("User does not have permission to update the field"))
+            except CustomProfileField.DoesNotExist:
+                raise JsonableError(
+                    _("Field id {id} not found.").format(id=custom_profile_field["id"])
+                )
+            except CustomProfileFieldValue.DoesNotExist:
+                pass
+
+
+def validate_user_can_remove_custom_profile_data(user_profile: UserProfile, field_id: int) -> None:
+    try:
+        field = CustomProfileField.objects.get(id=field_id)
+        if not field.editable_by_user and not user_profile.is_realm_admin:
+            raise JsonableError(_("User does not have permission to update the field"))
```

Comment:
In this codepath too, we query the database to get the same object twice. I think this check can also be moved to `check_remove_custom_profile_field_value` for now since we already query the database there to get the object.

Reply:
I have updated the PR now verification is done in `check_remove_custom_profile_field_value`.